cpp/src/text/utilities/tokenize_ops.cuh

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/string_view.cuh>
+
+#include <thrust/logical.h>
+
+namespace nvtext
+{
+namespace detail
+{
+
+using string_index_pair = thrust::pair<const char*,cudf::size_type>;
+using position_pair = thrust::pair<cudf::size_type,cudf::size_type>;
+
+/**
+ * @brief Tokenizer class that use multi-character delimiters.
+ *
+ * This is common code for tokenize, token-counters, normalize functions.
+ * If an empty delimiter string is specified, then whitespace
+ * (code-point <= ' ') is used to identify tokens.
+ *
+ * After instantiating this object, use the `next_token()` method
+ * to parse tokens and the `token_byte_positions()` to retrieve the
+ * current token's byte offsets within the string.
+ */
+struct characters_tokenizer
+{
+    __device__ characters_tokenizer( cudf::string_view const& d_str,
+                                     cudf::string_view const& d_delimiter = cudf::string_view{} )
+    : d_str(d_str), d_delimiter(d_delimiter), spaces(true), itr(d_str.begin()),
+      start_position(0), end_position(d_str.length()) {}
```

Comment:
Always use `{}` in the initialization list instead of `()`. The compiler silences narrowing conversions if you use `()`